src/simulators/statevector/qubitvector.hpp

```diff
@@ -845,160 +1230,600 @@ complex_t QubitVector<statevector_t>::apply_reduction_lambda(const std::vector<u
+  };
+  // Use the lambda function
+  apply_matrix_lambda(lambda, qs, mat);
+}
+
+template <class data_t>
+template <size_t N>
+void QubitVector<data_t>::apply_diagonal_matrix(const areg_t<N> &qs,
+                                                const cvector_t &diag) {
+
+  // Error checking
+  #ifdef DEBUG
+  check_vector(diag, N);
+  #endif
+
+  // Lambda function for N-qubit matrix multiplication
+  auto lambda = [&](const areg_t<1ULL << N> &inds,
+                    const cvector_t &_mat)->void {
+    const uint_t dim = 1ULL << N;
+    for (size_t i = 0; i < dim; i++) {
+      data_[inds[i]] *= _mat[i];
+    }
+  };
+
+  // Use the lambda function
+  apply_matrix_lambda(lambda, qs, diag);
+}
+
+// Static number of pairs
+template <class data_t>
+template <size_t N, size_t M>
+void QubitVector<data_t>::apply_permutation_matrix(const areg_t<N> &qubits,
+                                                   const std::array<std::pair<uint_t, uint_t>, M> &pairs) {
+  // Lambda function for permutation matrix
+  auto lambda = [&](const areg_t<1ULL << N> &inds)->void {
+    complex_t cache;
+    for (const auto& p : pairs) {
+      cache = data_[inds[p.first]];
+      data_[inds[p.first]] = data_[inds[p.second]];
+      data_[inds[p.second]] = cache;
+    }
+  };
+  // Use the lambda function
+  apply_lambda(lambda, qubits);
+}
+
+// Dynamic number of pairs
+template <class data_t>
+template <size_t N>
+void QubitVector<data_t>::apply_permutation_matrix(const areg_t<N> &qubits,
+                                                   const std::vector<std::pair<uint_t, uint_t>> &pairs) {
+  // Lambda function for permutation matrix
+  auto lambda = [&](const areg_t<1ULL << N> &inds)->void {
+    complex_t cache;
+    for (const auto& p : pairs) {
+      cache = data_[inds[p.first]];
+      data_[inds[p.first]] = data_[inds[p.second]];
+      data_[inds[p.second]] = cache;
+    }
+  };
+  // Use the lambda function
+  apply_lambda(lambda, qubits);
+}
+
+//------------------------------------------------------------------------------
+// Dynamic N
+//------------------------------------------------------------------------------
+
+/* Generate the repeatative switch cases using the following python code:
+
+```python
+code = ""
+for j in range(3, 16 + 1):
+    code += "  case {0}: {{\n".format(j)
+    code += "    areg_t<{0}> qubits_arr;\n".format(j)
+    code += "    std::copy_n(qubits.begin(), {0}, qubits_arr.begin());\n".format(j)
+    code += "    apply_matrix<{0}>(qubits_arr, mat);\n".format(j)
+    code += "  }} break;\n".format(j)
+print(code)
+```
+*/
+template <class data_t>
+void QubitVector<data_t>::apply_matrix(const reg_t &qubits,
+                                       const cvector_t &mat) {
+  // Special low N cases using faster static indexing
+  switch (qubits.size()) {
+  case 1:
+    apply_matrix(areg_t<1>({{qubits[0]}}), mat);
+    break;
+  case 2:
+    apply_matrix(areg_t<2>({{qubits[0], qubits[1]}}), mat);
+    break;
+  case 3: {
```

Comment:
We can compact the code a lot by passing-through from case 3 to case 16, like:
```c++
auto qubits_size = qubits.size();
switch(qubits_size) {
...
case 3:
case 4:
case 5:
...
case 16:
    areg_t<qubits_size> qubits_arr;
    std::copy_n(qubits.begin(), qubits_size, qubits_arr.begin());
    apply_matrix(qubits_arr, mat);
    break;
}
```

Reply:
I don't think this will compile because for the std::array methods the size needs to be known at compile time, hence why these big switch statements are needed.

Reply:
You are right. Nevermind :)